WalletWasabi/Services/UpdateManager.cs

```diff
@@ -0,0 +1,238 @@
+		using var stream = await httpClient.GetStreamAsync(url).ConfigureAwait(false);
+		Logger.LogInfo("Installer stream downloaded, copying...");
+		var file = File.OpenWrite(tmpFileName);
+		await stream.CopyToAsync(file).ConfigureAwait(false);
+
+		// Closing the file to rename.
+		file.Close();
+		File.Move(tmpFileName, newFileName);
+
+		InstallerName = fileName;
+
+		return true;
+	}
+
+	private (string url, string fileName) GetAssetToDownload(List<string> urls)
+	{
+		if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+		{
+			var url = urls.Where(url => url.Contains(".msi")).First();
+			return (url, url.Split("/").Last());
+		}
+		else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+		{
+			var cpu = RuntimeInformation.ProcessArchitecture;
+			if (cpu.ToString() == "Arm64")
+			{
+				var arm64url = urls.Where(url => url.Contains("arm64.dmg")).First();
+				return (arm64url, arm64url.Split("/").Last());
+			}
+			var url = urls.Where(url => url.Contains(".dmg") && !url.Contains("arm64")).First();
+			return (url, url.Split("/").Last());
+		}
+		else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+		{
+			throw new InvalidOperationException("For Linux, get the correct update manually.");
+		}
+		else
+		{
+			throw new InvalidOperationException("OS not recognized, download manually.");
+		}
+	}
+
+	public event EventHandler<UpdateStatus>? UpdateAvailableToGet;
+
+	public string DownloadsDir { get; }
+	public IHttpClient HttpClient { get; }
+
+	///<summary> Comes from config file. Decides Wasabi should download the new installer in the background or not.</summary>
+	public bool DownloadNewVersion { get; set; }
+
+	///<summary> Install new version on shutdown or not.</summary>
+	public bool DoUpdateOnClose { get; set; } = false;
+
+	public void InstallNewVersion()
+	{
+		try
+		{
+			var installerPath = Path.Combine(DownloadsDir, InstallerName);
+			ProcessStartInfo startInfo;
+			if (!File.Exists(installerPath))
+			{
+				throw new FileNotFoundException(installerPath);
+			}
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				startInfo = ProcessStartInfoFactory.Make(installerPath, "", true);
+			}
+			else
+			{
+				startInfo = new()
+				{
+					FileName = installerPath,
+					UseShellExecute = true,
+					WindowStyle = ProcessWindowStyle.Normal
+				};
+			}
+
+			using Process? p = Process.Start(startInfo);
+
+			if (p is null)
+			{
+				throw new InvalidOperationException($"Can't start {nameof(p)} {startInfo.FileName}.");
+			}
+			if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+			{
+				// For MacOS, you need to start the process twice, first start => permission denied
+				// TODO: find out why and fix.
+
+				p!.WaitForExit(5000);
+				p.Start();
+			}
+		}
+		catch (Exception ex)
+		{
+			Logger.LogError("Failed to install latest release. File might be corrupted.", ex);
+		}
+	}
+
+	public bool CheckIfInstallerDownloaded()
```

Comment:
This method does multiple things. Creating folders, deleting files, and looking for files. Try to keep the scope on CheckIfInstallerDownloaded, all the others should be done outside.